jina/drivers/encode.py

```diff
@@ -1,23 +1,117 @@
-    """Drivers inherited from this Driver will bind :meth:`craft` by default """
+    """Drivers inherited from this Driver will bind :meth:`encode` by default """
 
-    def __init__(self, executor: str = None, method: str = 'encode', *args, **kwargs):
+    def __init__(self,
+                 executor: str = None,
+                 method: str = 'encode',
+                 *args, **kwargs):
         super().__init__(executor, method, *args, **kwargs)
 
 
 class EncodeDriver(BaseEncodeDriver):
-    """Extract the chunk-level content from documents and call executor and do encoding
+    """Extract the content from documents and call executor and do encoding
     """
 
+    class CacheDocumentSet:
+        """Helper class to accumulate documents from differents document Set in a single DocumentSet
+         to help guarantee that the encoder driver can consume documents in fixed batch sizes to allow
+         the EncoderExecutors to leverage its batching abilities.
+
+         It is useful to have batching even when chunks are involved"""
+
+        def __init__(self,
+                     capacity: Optional[int] = None,
+                     *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.capacity = capacity
+            self._doc_set = DocumentSet(docs_proto=[])
+
+        @property
+        def available_capacity(self):
+            return self.capacity - len(self._doc_set)
+
+        def cache(self, docs: DocumentSet):
+            docs_to_append = min(len(docs), self.available_capacity)
+            for doc in docs[: docs_to_append]:
+                self._doc_set.append(doc)
+            return DocumentSet(docs[docs_to_append:])
+
+        def __len__(self):
+            return len(self._doc_set)
+
+        def get(self):
+            return self._doc_set
+
+    def __init__(self,
+                 batch_size: Optional[int] = None,
+                 *args, **kwargs):
+        """
+        Extract from different documents.
+        :param batch_size: number of documents to be used simultaniously in the encoder :meth:apply_all.
+        It is specially useful when the same EncoderExecutor can be used for documents of different granularities
+         (chunks, chunks of chunks ...)
+
+    .. warning::
+        - This parameter was added to cover the case where root documents had very few chunks, and the encoder executor could
+        then only process them in batches of the chunk size of each document, which did not lead to the full use of batching capabilities
+        of the powerful Executors
+
+        """
+        super().__init__(*args, **kwargs)
+        self.batch_size = batch_size
+        if self.batch_size:
+            self.cache_set = EncodeDriver.CacheDocumentSet(capacity=self.batch_size)
+        else:
+            self.cache_set = None
+
+    def __call__(self, *args, **kwargs):
+        self._traverse_apply(self.docs, *args, **kwargs)
+        self._apply_cache(force_flush=True, **kwargs)
```

Comment:
Because there may be some left-overs from the `request` (Imagine request batch size is 100 documents and driver_batch_size is 9) There will be 9 apply_all of batch_size 9 and then there will be 1 document leftover from the request that needs to be applied (1 that will remain in the cache). And it would not be safe to assume that another request would come